portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -23,10 +36,81 @@
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+    
+  List<String> commentsList = new ArrayList<String>();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    
+    // Create query for Datastore.
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // Iterate over results and add each comment to the ArrayList.
+    List<Comment> commentsList = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String username = (String) entity.getProperty("username");
+      String comments = (String) entity.getProperty("comments");
+      String timestamp = (String) entity.getProperty("timestamp");
+
+      Comment comment = new Comment(username, comments, timestamp);
+      commentsList.add(comment);
+    }
+
+    Gson gson = new Gson();
+
+    // Send the JSON as the response.
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(commentsList));
+  }
+  
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String username = getParameter(request, "username", "");
+    String comments = getParameter(request, "comments", "");
+    SimpleDateFormat formatter = new SimpleDateFormat("EEE dd/MM/yyyy hh:mm z");  
+    Date date = new Date();  
+    String timestamp = formatter.format(date).toString();  
+
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("username", username);
+    commentEntity.setProperty("comments", comments);
+    commentEntity.setProperty("timestamp", timestamp);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter
+   *         was not specified by the client
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  /**
+   * Converts the form input into a JSON string using manual String concatentation.
+   */
+  private String convertCommentToJson(String username, String comments) {
+    String json = "{";
```

Comment:
I think you can replace this with `String.format("{ username: %s, comments: %s }", username, comments);`. The extra quotation marks shouldn't be necessary :)

Reply:
wow thanks for this!